cocotb/share/lib/vpi/VpiCbHdl.cpp

```diff
@@ -49,77 +49,85 @@ VpiCbHdl::VpiCbHdl(GpiImplInterface *impl) : GpiCbHdl(impl)
     cb_data.user_data = (char*)this;
 }
 
-/* If the user data already has a callback handle then deregister
- * before getting the new one
- */
 int VpiCbHdl::arm_callback() {
+    switch(m_state) {
+    case GPI_PRIMED:
+    case GPI_CALL:
+        LOG_WARN("VPI: Attempted to prime an already primed trigger for %s (%d)",
+                  m_impl->reason_to_string(cb_data.reason), cb_data.reason);
+        /* FALLTHRU */
+    case GPI_FREE: {
+        if (m_obj_hdl != NULL) {
+            if (m_state == GPI_FREE) {
+                // LCOV_EXCL_START
+                LOG_ERROR("VPI: Simulator callback already registered for %s (%d)",
+                        m_impl->reason_to_string(cb_data.reason), cb_data.reason);
+                return -1;
+                // LCOV_EXCL_STOP
+            } else {
+                cleanup_callback();
+            }
+        }
 
-    if (m_state == GPI_PRIMED) {
-        fprintf(stderr,
-                "Attempt to prime an already primed trigger for %s!\n",
-                m_impl->reason_to_string(cb_data.reason));
-    }
+        vpiHandle new_hdl = vpi_register_cb(&cb_data);
+        check_vpi_error();
 
-    // Only a problem if we have not been asked to deregister and register
-    // in the same simulation callback
-    if (m_obj_hdl != NULL && m_state != GPI_DELETE) {
-        fprintf(stderr,
-                "We seem to already be registered, deregistering %s!\n",
-                m_impl->reason_to_string(cb_data.reason));
-        cleanup_callback();
+        if (!new_hdl) {
+            // LCOV_EXCL_START
+            LOG_ERROR("VPI: Unable to register a simulator callback for %s (%d)",
+                      m_impl->reason_to_string(cb_data.reason), cb_data.reason);
+            return -1;
+            // LCOV_EXCL_STOP
+        }
+        set_call_state(GPI_PRIMED);
+        m_obj_hdl = new_hdl;
+        break;
     }
 
-    vpiHandle new_hdl = vpi_register_cb(&cb_data);
+    case GPI_DELETE:
+        // This state is used by VpiTimedCbHdl only
+        LOG_DEBUG("VPI: Re-priming callback for %s (%d)", m_impl->reason_to_string(cb_data.reason), cb_data.reason);
+        set_call_state(GPI_PRIMED);
+        break;
```

Comment:
This can't be tested from Python because the `Timer.cbhdl` (handle to this object) is dropped in `unprime()`.